### PR TITLE
Move teamd warm reboot code to service script

### PR DESCRIPTION
**Pull request by heidinet2007**
Move teamd warm reboot code to teamd service at /usr/local/bin

### CODE REVIEW

**Comment by heidinet2007** [comment]
please re-test

**Comment by heidinet2007** [comment]
please retest

File: files/scripts/teamd.sh [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    debug "Fast boot flag: ${SERVICE}$DEV ${FAST_BOOT}."
+
+    if [[ x"$WARM_BOOT" == x"true" ]] || [[ x"$FAST_BOOT" == x"true" ]]; then
+		# Kill teamd processes inside of teamd container with SIGUSR2 to allow them to send last LACP frames
+		# We call `docker kill teamd` to ensure the container stops as quickly as possible,
+		# then immediately call `systemctl stop teamd` to prevent the service from
+		# restarting the container automatically.
+		# Note: teamd must be killed before syncd, because it will send the last packet through CPU port
+		debug "Stopping teamd ..."
+		docker exec -i teamd pkill -USR2 teamd || [ $? == 1 ]
```
Comment (yxieca) [inline files/scripts/teamd.sh:76]: I think fast/warm reboot are using different signals USR1 v.s. USR2.

File: files/scripts/teamd.sh [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        # We call `docker kill teamd` to ensure the container stops as quickly as possible,
+        # then immediately call `systemctl stop teamd` to prevent the service from
+        # restarting the container automatically.
+        # Note: teamd must be killed before syncd, because it will send the last packet through CPU port
+        docker exec -i ${SERVICE}$DEV pkill -USR2 ${SERVICE} || [ $? == 1 ]
+        while docker exec -i ${SERVICE}$DEV pgrep ${SERVICE} > /dev/null; do
+            sleep 0.05
+        done
+        docker kill ${SERVICE}$DEV  &> /dev/null || debug "Docker ${SERVICE}$DEV is not running ($?) ..."
+        systemctl stop ${SERVICE}$DEV
```
Comment (yxieca) [inline files/scripts/teamd.sh:86]: Let's remove this systemctl stop, it feels like a circular dependency.

File: files/scripts/teamd.sh [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    else
+        /usr/bin/${SERVICE}.sh stop $DEV
+    fi
+
+    debug "Stopped ${SERVICE}$DEV service..."
```
Comment (yxieca) [inline files/scripts/teamd.sh:91]: Let's move this line outside if / elif, and remove the else clause.

File: files/scripts/teamd.sh [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        # restarting the container automatically.
+        # Note: teamd must be killed before syncd, because it will send the last packet through CPU port
+        docker exec -i ${SERVICE}$DEV pkill -USR2 ${SERVICE} || [ $? == 1 ]
+        while docker exec -i ${SERVICE}$DEV pgrep ${SERVICE} > /dev/null; do
+            sleep 0.05
+        done
+        docker kill ${SERVICE}$DEV  &> /dev/null || debug "Docker ${SERVICE}$DEV is not running ($?) ..."
+        systemctl stop ${SERVICE}$DEV
+    else
+        /usr/bin/${SERVICE}.sh stop $DEV
```
Comment (yxieca) [inline files/scripts/teamd.sh:88]: Let's move this line outside the if-elif block and remove the else clause.

File: files/scripts/teamd.sh (hunk 2 above)
Comment (heidinet2007) [reply files/scripts/teamd.sh]: Looks like this line was added in https://github.com/Azure/sonic-utilities/pull/572. We don't need it any more?

**Review by yxieca — DISMISSED** [review]

File: files/scripts/teamd.sh (hunk 2 above)
Comment (yxieca) [reply files/scripts/teamd.sh]: As we discussed offline. This line could cause recursive service stop.

File: files/build_templates/sonic_debian_extension.j2 [hunk 5]
```diff
@@ -533,6 +533,7 @@ sudo LANG=C chroot $FILESYSTEM_ROOT umount -lf /sys
 sudo LANG=C cp $SCRIPTS_DIR/swss.sh $FILESYSTEM_ROOT/usr/local/bin/swss.sh
 sudo LANG=C cp $SCRIPTS_DIR/syncd.sh $FILESYSTEM_ROOT/usr/local/bin/syncd.sh
 sudo LANG=C cp $SCRIPTS_DIR/bgp.sh $FILESYSTEM_ROOT/usr/local/bin/bgp.sh
+sudo LANG=C cp $SCRIPTS_DIR/teamd.sh $FILESYSTEM_ROOT/usr/local/bin/teamd.sh
```
Comment (vaibhavhd) [inline files/build_templates/sonic_debian_extension.j2:536]: The comment above can be updated too.

File: files/scripts/teamd.sh [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        FAST_BOOT="true"
+    else
+        FAST_BOOT="false"
+    fi
+}
+
+start() {
+    debug "Starting ${SERVICE}$DEV service..."
+
+    check_warm_boot
```
Comment (vaibhavhd) [inline files/scripts/teamd.sh:43]: Should we also wait for database initialization before accessing it? I see that some of the service scripts use it, and some are missing it already.

https://github.com/Azure/sonic-buildimage/blob/master/files/scripts/swss.sh#L60
https://github.com/Azure/sonic-buildimage/blob/master/files/scripts/swss.sh#L137

**Review by vaibhavhd — COMMENTED** [review]
Minor comments added. It would be also better if we have output logs for `start,stop,wait`.

**Comment by yxieca** [comment]
retest broadcom please

**Comment by yxieca** [comment]
retest vsimage please

**Comment by yxieca** [comment]
retest vsimage please

**Comment by yxieca** [comment]
retest vsimage please

File: files/build_templates/sonic_debian_extension.j2 (hunk 5 above)
Comment (heidinet2007) [reply files/build_templates/sonic_debian_extension.j2:536]: Updated.

File: files/scripts/teamd.sh [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        WARM_BOOT="true"
+    else
+        WARM_BOOT="false"
+    fi
+}
+
+function validate_restore_count()
+{
+    if [[ x"$WARM_BOOT" == x"true" ]]; then
+        RESTORE_COUNT=`$SONIC_DB_CLI STATE_DB hget "WARM_RESTART_TABLE|teamd" restore_count`
```
[the remark this answers is not in the document]
Comment (heidinet2007) [reply files/scripts/teamd.sh]: changed it to ${SERVICE}

File: files/scripts/teamd.sh (hunk 6 above)
Comment (heidinet2007) [reply files/scripts/teamd.sh:43]: I saw this wait_for_database_service() is called from swss and syncd.  Teamd already has dependancy on swss.

File: files/scripts/teamd.sh [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+    check_fast_boot
+
+    debug "Warm boot flag: ${SERVICE}$DEV ${WARM_BOOT}."
+    debug "Fast boot flag: ${SERVICE}$DEV ${Fast_BOOT}."
+
+    # start service docker
+    /usr/bin/${SERVICE}.sh start $DEV
+    debug "Started ${SERVICE}$DEV service..."
+
```
[the remark this answers is not in the document]
Comment (heidinet2007) [reply files/scripts/teamd.sh]: Removed

File: files/scripts/teamd.sh [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+DEV=$2
+
+SERVICE="teamd"
+DEBUGLOG="/tmp/teamd-debug$DEV.log"
+NAMESPACE_PREFIX="asic"
+if [ "$DEV" ]; then
+    NET_NS="$NAMESPACE_PREFIX$DEV" #name of the network namespace
+    SONIC_DB_CLI="sonic-db-cli -n $NET_NS"
+else
+    NET_NS=""
```
[the remark this answers is not in the document]
Comment (heidinet2007) [reply files/scripts/teamd.sh]: Removed

**Review by yxieca — DISMISSED** [review]

**Review by vaibhavhd — DISMISSED** [review]
LGTM, thanks Heidi.

**Comment by vaibhavhd** [comment]
Retest vsimage please

**Comment by vaibhavhd** [comment]
Retest mellanox please

**Comment by vaibhavhd** [comment]
Retest vsimage please

**Comment by vaibhavhd** [comment]
Retest vsimage please

**Comment by vaibhavhd** [comment]
Retest vsimage please

**Comment by vaibhavhd** [comment]
Retest vsimage please

**Comment by yxieca** [comment]
retest vsimage please

**Comment by vaibhavhd** [comment]
The vsimage tests are failing due to incorrect permissions set on a new teamd script being added by this PR. Below is the fix:
```
$ git diff
diff --git a/files/scripts/teamd.sh b/files/scripts/teamd.sh
old mode 100644
new mode 100755
```
This would prevent the below error seen in syslog:
```
Nov  9 18:45:02.563249 sonic INFO systemd[1]: Starting TEAMD container...
Nov  9 18:45:02.564008 sonic ERR systemd[1867]: teamd.service: Failed to execute command: Permission denied
Nov  9 18:45:02.564329 sonic ERR systemd[1867]: teamd.service: Failed at step EXEC spawning /usr/local/bin/teamd.sh: Permission denied
```

@heidinet2007 I am unable to make this change on the fork. Can you make this change, or give me permission to do so?

**Comment by yxieca** [comment]
> The vsimage tests are failing due to incorrect permissions set on a new teamd script being added by this PR. Below is the fix:
> 
> ```
> $ git diff
> diff --git a/files/scripts/teamd.sh b/files/scripts/teamd.sh
> old mode 100644
> new mode 100755
> ```
> 
> This would prevent the below error seen in syslog:
> 
> ```
> Nov  9 18:45:02.563249 sonic INFO systemd[1]: Starting TEAMD container...
> Nov  9 18:45:02.564008 sonic ERR systemd[1867]: teamd.service: Failed to execute command: Permission denied
> Nov  9 18:45:02.564329 sonic ERR systemd[1867]: teamd.service: Failed at step EXEC spawning /usr/local/bin/teamd.sh: Permission denied
> ```
> 
> @heidinet2007 I am unable to make this change on the fork. Can you make this change, or give me permission to do so?

@heidinet2007 I've pushed a change to your branch to change file permission.

**Comment by yxieca** [comment]
retest vsimage please

**Review by vaibhavhd — APPROVED** [review]
LGTM, thanks!

**Pull request merged**